Vue migration: settings

skjnldsv opened the pull request:
Features:
- Form validation
  - New user form requires user+pass or user+email minimum
  - Validate every other fields:
    - displayname can be empty
    - password can't be empty and has the minlength requirement from the password policy if enabled
    - emails are validated as emails & can be empty
- We switched to a grid design, no more ugly table
- Hide/show columns settings are now saved in the browser, saving us some undesirable db entries
- Only using our api, no more shady legacy ajax entry points
- We only show 2 groups max, and [add an indicator](https://user-images.githubusercontent.com/14975046/37727800-03f2e09c-2d39-11e8-9ee1-d45daf531071.png) saying how many more are selected (better visual & consistency)
- Password can't be change if encryption does NOT use a master key (safety reason)
- Now using Vuejs, components are fresh and will be moved to a dedicated repo later on with proper documentation and tests
- Language edition capability

Fix #642 
Fix #7424
Fix #5752 
Fix #5751
Fix #8862
Fix #9289
Fix #9191
Fix #8917 
Fix #8010
Fix #4320

Settings
- [x] Design + base
- [x] Quota
- [x] Groups management
- [x] Subadmins management
- [x] User disablement
- [x] User deletion
- [x] Props edition (email, name...)
- [x] Server-side-ecrptn password change
- [x] Migrate to OCS api
- [x] Support password policy min length
- [ ] Search support
- [x] Group deletion

Design
- [x] Better display of overflow tags
- [x] Themed vue-multiselect

Waiting for:
- [x] #8856 
- [x] #8847 
- [x] #8864 
- [x] #8865
- [x] #8904 
- [x] #8926
- [x] #9111 
- [x] #8777
- [x] #9108 
- [x] #9296 
- [ ] #9438

Components see #7738 
- [x] Popovermenu v1
- [x] AppNavigation v1
- [x] Request api + authentification system

Bugs
- [ ] On user deletion, if admin of group(s), db key(s) is/are not removed
- [x] All user count also includes disabled but they are not shown on the list
- [ ] User old apppassword are not deleted on user deletion! :warning: 

Before merge:
- [x] Drop css fixed commits after #8777 
- [x] Drop 582aece610d0d98852ee1ddf0f8f6a48e63c14cb after #9108 is merged
- [x] Drop a361b7b038c33b8215f369b199fbcc8940c00539 after  #9111 is merged

skjnldsv committed:
Settings to vuejs

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

MorrisJobke commented on settings/README.md:
```diff
@@ -0,0 +1,18 @@
+npm run dev
+
+# build for production with minification
+npm run build
```
Makefile please ... see `apps/updaternotifications` 😉

skjnldsv replied:
I'll implement it once done :)

nickvergessen commented on settings/users.php:
```diff
@@ -109,7 +110,7 @@
 $disabledUsersGroup = [
 	'id' => '_disabledUsers',
-	'name' => '_disabledUsers',
+	'name' => 'Disabled users',
```
translate?!

skjnldsv committed:
Group filtering, disable state, ocs api final update

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv committed:
Makefile

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv committed:
Redirect if disabled list is empty

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv committed:
Added language support in users list

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv committed:
Fixed tabs and Language support

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv committed:
Loading state to vue and not css

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv committed:
Settings users list new user language

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv committed:
Various fixes and group deletion

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

nickvergessen committed:
Allow icon urls too

Signed-off-by: Joas Schilling <coding@schilljs.com>

nickvergessen commented:
- [ ] Loading icon/state for the items-list

skjnldsv committed:
Supports app navigation loading state

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv committed:
Bump user controller, cleaned old legacy and added route history mode

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv committed:
Continued the upgrade to proper router link and added navigationItem router linkTo capability

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv committed:
Added default quota selector

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

nickvergessen commented on settings/src/components/appNavigation/navigationItem.vue:
```diff
@@ -92,18 +92,18 @@ export default {
 	},
 	methods: {
-		showMenu () {
+		showMen() {
```
?

skjnldsv replied:
Error indeed

skjnldsv committed:
new OC api and default quota fixes

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv committed:
Fixed settings menu entry and group deletion

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv committed:
Last login tooltip and v-tooltip server compatibility css

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv committed:
Bump makefile, readme, fix typo and cleanup

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv committed:
Bump popovermenu item options and tests

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv committed:
Bump tests and editorconfig

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv committed:
Fix Applicationtest

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv committed:
Fix row detection acceptance test

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv committed:
Composer autoloader bump

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv committed:
Fixed new user form detection in acceptance

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv committed:
Fixed router

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv committed:
newt acceptance tests

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv committed:
Users list acceptance tests

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv committed:
Tabs fix

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv commented:
Okay everyone, this is ready for review!
This is unfortunately a pretty big PR, because I cleaned a lot of old legacy stuff. Please don't be afraid! <3

@nextcloud/designers 

What to test:
- Addition/removal of users
- Addition/removal of groups
- Quota change
- Default quota change
- Password/displayname/email changes
- Groups/subadmins edition
- Custom quotas set/unset
- Group change on the app navigation
- Load the page with http://xxxx/index.php/xxx and without `index.php`
- Language change on users
- Password policy min length requirements

skjnldsv committed:
Bump script and fixed copyright template

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv committed:
Add feature to drone

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

juliushaertl commented:
Really nice so far, just a few things from having a first try, i'll have a more detailed look tomorrow:

- When index.php is required for the URL the router doesn't seem to add it when clicking on groups in the navigation sidebar
- Clicking the 3-dots menu in the user list for the second time doesn't close the popover again
- The delete button for groups should be hidden by default in my opinion and just be shown on hover (or maybe as it was before just if you selected the group)
- A bit more space between the columns would be good for when inputs are besides plain text 
![image](https://user-images.githubusercontent.com/3404133/39874922-72f6d324-546f-11e8-82ff-42614cd6b2b5.png)
- The group multiselect input should have a pointer cursor to indicate that it is clickable
- Text fields and multiselect fields should have the same height
![image](https://user-images.githubusercontent.com/3404133/39874969-9ea58f9c-546f-11e8-86ea-2bb7edbe2f06.png)
- When clicking the new user button, the first input field should be focused

skjnldsv commented:
> The delete button for groups should be hidden by default in my opinion and just be shown on hover (or maybe as it was before just if you selected the group)

I'm not sure on this. I agree that it looks strange, but it's also a bad UX for users, like you don't know where you can delete the group. We could try something else tough, but I don't have any idea on this atm :(

Thanks for your review, I'll fix this asap :heart:

skjnldsv committed:
Design fixes
- Gap between columns
- Click on icon more closes the menu
- Fixed select height
- Focus input on new user click
- Cursor fix on select

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv committed:
Production build, router base url fix

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

juliushaertl commented on settings/src/router.js:
```diff
@@ -15,6 +15,9 @@ Vue.use(Router);
 	mode: 'history',
+	// if index.php is in the url AND we got this far, then it's working:
+	// let's keep using index.php in the url
+	base: window.location.pathname.indexOf('index.php') > 0 ? '/index.php/' : '/',
```
I think that would still break when installed into a subdirectory. Can we just use `OC.generateUrl('')`?

skjnldsv replied:
Good call. I need to determine the base url yes :)

skjnldsv committed:
Fixed root url detection, new tests, default new user group to current
opened group and user removal update fix

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

juliusknorr reviewed (dismissed):
Looks good code wise, just two bugs with groups management found.

juliushaertl commented:
Two other issues with groups:
- [ ] Add a user while you are viewing a group list -> User is added without a group, imo we should prefill the group list with the current group there
- [ ] If you unassign a user from the group you are currently viewing it should be removed from the list

skjnldsv commented on lib/private/L10N/Factory.php:
```diff
@@ -441,4 +446,74 @@ public function createPluralFunction($string) {
+	 *
+	 * @return array
+	 */
+	public function getLanguages() {
```
For the sanity of reviewers, I guess tests for this will come later ^^

MorrisJobke commented:
> We switched to a grid design, no more ugly table

How does it look in IE?

skjnldsv committed:
IE11 compatibility

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv commented:
> How does it look in IE?

Descent for ie11!
Some center align issues, but nothing terrible. I added a config for ie polyfills. 
In the end we should really drop ie11!

michag86 commented:
Hi,
thanks for the work on the user management!

I have tested it and found some issues:
- [ ] Groups with long names are cut off in the group selection of the user.
- [ ] It would be nice if the results are sortable
- [ ] Problems for subadmins (/ocs/v2.php/apps/provisioning_api/api/v1/config/apps/password_policy/minLength) results in error 997 "Logged in user must be an admin"
- [ ] If subadmin trys to create a user without selecting a group, the spinner is spinning for ever. 
- [ ] Subadmin can see the subadmin permissions and can try to change subadmin permissions. Maybe this should be hidden, hideable or read only for subadmins?
- [ ] User count for everyone (as subadmin) is not the count of the subadmin scope. Instead it represents all users.
- [ ] Subadmins can try to create a group and the spinner is spinning for ever.

skjnldsv commented:
> - It would be nice if the results are sortable

Maybe in a future pr :)

> - Problems for subadmins (/ocs/v2.php/apps/provisioning_api/api/v1/config/apps/password_policy/minLength) results in error 997 "Logged in user must be an admin"

Have to be fixed in the policy app: https://github.com/nextcloud/password_policy

skjnldsv committed:
Various fixes
* Fixed  subadmin group management
* Fixed pass policy prop retrieving
* Fixed fields subadmins were not supposed to see

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv committed:
Proper error message and group order fix in select

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv commented:
@nextcloud/designers bump :)

skjnldsv committed:
Bump webpack, config, deps, fixes groups selects, improved design and added tooltip to groups +x indicator

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv commented:
@michag86 small push to update the width of the groups list
Also: added a tooltip to show the full list on hover: 
![capture d ecran_2018-05-18_16-58-20](https://user-images.githubusercontent.com/14975046/40242003-af2ce24a-5abc-11e8-9755-64b9a32f725c.png)

skjnldsv committed:
Fixed webpack config, groups flex thanks to @juliushaertl and groups tooltip

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv committed:
Fix api error

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

MorrisJobke reviewed (approved):
Tested and works a lot better than before - some minor things for Edge and IE, but this is stuff for later 👍

rullzer reviewed (approved):
:elephant:

danopz commented:
Will the components be pushed to [nextcloud/vue-components](https://github.com/nextcloud/vue-components)?

skjnldsv commented:
@danopz very soon yes :)
